File: app/src/main/java/com/example/chadrick/datalabeling/Models/LabelDrawPad.java
package com.example.chadrick.datalabeling.Models;

import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Point;
import android.graphics.Rect;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.ImageView;
import com.example.chadrick.datalabeling.Callback;
import com.example.chadrick.datalabeling.CallbackWithRect;
import com.example.chadrick.datalabeling.CustomComponents.BasicZoomImageView;
import com.example.chadrick.datalabeling.CustomComponents.MaskImageView;
import com.example.chadrick.datalabeling.CustomComponents.PageFrameLayout2;
import com.example.chadrick.datalabeling.CustomComponents.TouchImageView;
import com.example.chadrick.datalabeling.CustomViewPager;
import com.example.chadrick.datalabeling.R;
import com.example.chadrick.datalabeling.Util;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Created by chadrick on 17. 11. 4.
 */

public class LabelDrawPad {

  private LayoutInflater inflater;
  private ViewGroup container;
  private int position;
  private View rootview;

  private BasicZoomImageView baseIV;
  private TouchImageView rectIV;
  private MaskImageView drawIV;
  private Callback drawBtnpressedcallback;
  private CallbackWithRect maskRectReadyCallback;

  private CustomViewPager customViewPager;
  private Paint mainCanvasPaint;

  private File imageFile, labelFile;
  private Bitmap baseBitmap, rectBitmap;
  private Canvas rectCanvas;
  private Rect selectedRect;
  private Runnable rectSelectedCallback;
  private Runnable hideDeleteBtnCallback;

  private ArrayList<Rect> rectArrayList = new ArrayList<Rect>();

  private static int RECT_SELECT_AREA_PADDING = 50;


  private final String TAG = this.getClass().getSimpleName();


  private void initElements() {
    rootview = inflater.inflate(R.layout.labeldrawpad_layout, container,
        false);

    baseIV = (BasicZoomImageView) rootview.findViewById(R.id.baseiv);
    rectIV = (TouchImageView) rootview.findViewById(R.id.rectiv);
    drawIV = (MaskImageView) rootview.findViewById(R.id.drawiv);

    // need to register the IVs so that it can receive touch events
    PageFrameLayout2 pageframelayout2 = (PageFrameLayout2) rootview
        .findViewById(R.id.pageframelayout2);
    pageframelayout2.registerIVs(baseIV, drawIV, rectIV);

    // create paint for maincanvas
    mainCanvasPaint = new Paint();
    mainCanvasPaint.setColor(Color.rgb(255, 63, 20));
    mainCanvasPaint.setStrokeWidth(5);
    mainCanvasPaint.setStyle(Paint.Style.STROKE);

    // dynamic width/height measure for drawIV
    ViewTreeObserver viewTreeObserver = drawIV.getViewTreeObserver();

    viewTreeObserver.addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
      @Override
      public void onGlobalLayout() {

        drawIV.passWH(drawIV.getWidth(), drawIV.getHeight());

        drawIV.getViewTreeObserver().removeOnGlobalLayoutListener(this);

      }
    });

    // create bitmap with imageFile and set up baseCanvas
    BitmapFactory.Options options = new BitmapFactory.Options();
    options.inPreferredConfig = Bitmap.Config.ARGB_8888;
    options.inMutable = true;

    Log.d(TAG, "initElements: imageFile=" + imageFile.toString());
    Log.d(TAG, "initElements: imageFile exist=" + imageFile.exists());
    Log.d(TAG, "initElements: imageFile size=" + imageFile.length());
    Bitmap originalbitmap = BitmapFactory.decodeFile(imageFile.toString(), options);
    baseBitmap = Bitmap.createBitmap(originalbitmap.getWidth(), originalbitmap.getHeight(),
        Bitmap.Config.ARGB_8888);
    Canvas basecanvas = new Canvas(baseBitmap);
    basecanvas.drawBitmap(originalbitmap, 0, 0, null);

    // create blank bitmap
    rectBitmap = Bitmap
        .createBitmap(originalbitmap.getWidth(), originalbitmap.getHeight(), Config.ARGB_8888);
    rectCanvas = new Canvas(rectBitmap);

    // set baseIV
    baseIV.setImageBitmap(baseBitmap);
    baseIV.setdrawBtnpressedcallback(drawBtnpressedcallback);

    // set rectIV
    rectIV.setImageBitmap(rectBitmap);
    rectIV.passCanvas(rectCanvas);
    rectIV.setCustomViewPager(
        customViewPager); // need this since swipe dis/enable will be controlled here
    rectIV.setdrawBtnpressedcallback(drawBtnpressedcallback);
    rectIV.passAddRectCallback((rect) -> {
      rectArrayList.add(rect);
    });
    rectIV.passSaveLabelCallback(this::saveLabelFile);
    rectIV.passCheckRectSelectCallback(this::checkRectSelect);
    rectIV.passUnselectAnyIfExist(this::unselectAnyIfExist);

    // drawIV has its own canvas
    drawIV.setdrawBtnpressedcallback(drawBtnpressedcallback);
    drawIV.passRectReadyCallback(maskRectReadyCallback);

    // generate label json file based on image file
    String imagefilename = Util.getOnlyFilename(imageFile.getName());
    String parentpath = imageFile.getParent();

    String printstr = parentpath + File.separator + imagefilename + ".json";
    Log.d(TAG, "json file : " + printstr);

    labelFile = new File(printstr);

    // check if labelfile exist. if exist, read it and draw them. if not exist, then do nothing.
    if (!labelFile.exists()) {

      // when json file doesn't exist
      Log.d(TAG, " label file(" + labelFile.toString() + "), doesn't exist");

    } else {
      // if json file exists, then read it into a json object
      Log.d(TAG, "json file exist. start parsing.");

      Log.d(TAG, "labelfile size" + labelFile.length());

      StringBuilder readtext = new StringBuilder();
      try {
        BufferedReader bufferedReader = new BufferedReader(new FileReader(labelFile));
        String line;
        while ((line = bufferedReader.readLine()) != null) {
          readtext.append(line);
        }

      } catch (IOException e) {
        e.printStackTrace();

      }

      String readresult = readtext.toString();
      Log.d(TAG, "read result: " + readresult);

      try {
        JSONObject parsed = new JSONObject(readresult);
        String bitmapfilename = parsed.getString("img");
        JSONArray rects = parsed.getJSONArray("rects");

        // for all the rects in the jsonfile, parse it and save it in the rectarraylist.
        if (rects.length() > 0) {
          // rects item exists. transfer them to rectarray
          for (int i = 0; i < rects.length(); i++) {
            JSONObject rectitem = rects.getJSONObject(i);
            int x1 = rectitem.getInt("x1");
            int y1 = rectitem.getInt("y1");
            int x2 = rectitem.getInt("x2");
            int y2 = rectitem.getInt("y2");

            Rect rectfromjson = new Rect(x1, y1, x2, y2);
            rectArrayList.add(rectfromjson);

          }
        }

      } catch (JSONException e) {
        e.printStackTrace();
      }

      // parsing the jsonfile finished.
      // now draw each rect in the maincanvas

      for (int i = 0; i < rectArrayList.size(); i++) {
        Rect rectToDraw = rectArrayList.get(i);
        rectCanvas.drawRect(rectToDraw, mainCanvasPaint);
      }

      // invalidate to make sure that the new drawing are displayed
      rectIV.invalidate();


    }

    Log.d(TAG, "initElements: finished");
  }

  // this is required to return the view to instantiating in fullimageadapter
  public View getRootview() {
    Log.d(TAG, "getRootview: returning rootview");
    return rootview;
  }

  private LabelDrawPad(LabelDrawPadBuilder builder) {
    this.inflater = builder.inflater;
    this.container = builder.container;
    this.position = builder.position;
    this.drawBtnpressedcallback = builder.drawBtnpressedcallback;
    this.maskRectReadyCallback = builder.maskRectReadyCallback;
    this.imageFile = builder.imageFile;
    this.customViewPager = builder.customViewPager;
    this.rectSelectedCallback = builder.rectSelectedCallback;
    this.hideDeleteBtnCallback = builder.hideDeleteBtnCallback;

    initElements();

  }


  public boolean saveLabelFile() {
    // when called, rewrite the labelfile with current info

    // first create a jsonobject and populate with current info
    JSONObject newRoot = new JSONObject();

    try {
      newRoot.put("img", imageFile.getName());

      // create JSONarray object that contains all the current rects
      JSONArray rectarray = new JSONArray();

      for (int i = 0; i < rectArrayList.size(); i++) {
        Rect rect = rectArrayList.get(i);
        JSONObject jsonrect = new JSONObject();
        jsonrect.put("x1", rect.left);
        jsonrect.put("y1", rect.top);
        jsonrect.put("x2", rect.right);
        jsonrect.put("y2", rect.bottom);

        rectarray.put(jsonrect);
      }

      newRoot.put("rects", rectarray);

    } catch (JSONException e) {
      e.printStackTrace();
      return false;
    }

    // newroot is ready. write it to labelFile

    try {
      BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(labelFile));
      bufferedWriter.write(newRoot.toString());
      bufferedWriter.flush();
      bufferedWriter.close();
      Log.d(TAG, "newly written contents:" + newRoot.toString());
      Log.d(TAG, "new labelfile size: " + labelFile.length());
    } catch (IOException e) {
      e.printStackTrace();
      return false;
    }
    // finished overwriting the labelfile
    // job finished
    // return true since we have successfully written it.
    return true;

  }


  public void enableTouches() {
    baseIV.setTouchEnable(true);
    rectIV.setTouchEnable(true);
    drawIV.setTouchEnable(true);
  }

  public void disableTouches() {
    baseIV.setTouchEnable(false);
    rectIV.setTouchEnable(false);
    drawIV.setTouchEnable(false);
  }

  private boolean checkRectSelect(Point point) {
    // check with all rects. the first hit will be set as selected and
    // this method will return true
    // if there are no hits, then return false

    int touch_x = point.x;
    int touch_y = point.y;

    Log.d(TAG, "checkRectSelect: received x=" + touch_x + ", y=" + touch_y);

    for (Rect rect : rectArrayList) {

      Rect biggerRect = new Rect(rect.left - RECT_SELECT_AREA_PADDING,
          rect.top - RECT_SELECT_AREA_PADDING,
          rect.right + RECT_SELECT_AREA_PADDING,
          rect.bottom + RECT_SELECT_AREA_PADDING);

      Rect smallerRect = new Rect(rect.left + RECT_SELECT_AREA_PADDING,
          rect.top + RECT_SELECT_AREA_PADDING,
          rect.right - RECT_SELECT_AREA_PADDING,
          rect.bottom - RECT_SELECT_AREA_PADDING);

      Log.d(TAG,
          "checkRectSelect: biggerRect, left=" + biggerRect.left + ", top=" + biggerRect.top + ","
              + "right=" + biggerRect.right + ", bottom=" + biggerRect.bottom);

      Log.d(TAG,
          "checkRectSelect: smallerREct, left=" + smallerRect.left + ", top=" + smallerRect.top +
              ", right=" + smallerRect.right + ", bottom=" + smallerRect.bottom);

      // assume that there are no conflicts with the sizes of the bigger and smaller rect

      boolean biggercontain = false;
      boolean smallercontain = false;

      if (biggerRect.contains(touch_x, touch_y)) {
        biggercontain = true;
      }

      if (smallerRect.contains(touch_x, touch_y)) {
        smallercontain = true;
      }

      Log.d(TAG,
          "checkRectSelect: biggercontain=" + biggercontain + ", smallercontain=" + smallercontain);

      if (biggercontain == true && smallercontain == false) {
        Log.d(TAG, "checkRectSelect: select rect hit found");

        // okay so a hit was found, and we know which rectangle it is.
        // we need to redraw it, and it will eventually be done with
        // rectIV
        // the rectangle of interest is already drawn.
        // instead of redrawing the whole thing(including othe rectangles),
        // why not just overdraw the rectangle of interest with anther paint?

        // first draw unselected rect of the previous existing selectedRect
        if (selectedRect != null) {
          rectIV.drawUnselectedRect(selectedRect);
        }

        selectedRect = rect;

        rectIV.drawSelectedRect(selectedRect);

        // display delete btn
        rectSelectedCallback.run();

        return true;
      }


    }

    return false;
  }

  public void drawRect() {
    rectIV.drawRect();
  }

  public void eraseDrawnRect() {
    drawIV.eraseall();
  }


  /**
   * this will delete the selectedRect from the list
   * and redraw canvas without that rect
   */
  public void deleteSelectedRect() {
    // delete the selectedrect

    if (selectedRect == null) {
      return;
    }

//    rectIV.drawDeleteRect(selectedRect);

    // remove it from rectarray
    Log.d(TAG, "deleteSelectedRect: length of rectarray before remove=" + rectArrayList.size());
    rectArrayList.remove(selectedRect);
    Log.d(TAG, "deleteSelectedRect: length of rectarray after remove=" + rectArrayList.size());

    // nullify the selectedrect variable
    selectedRect = null;

//    // clear the rectIV, and redraw all rects in rectarray
//    rectBitmap.eraseColor(Color.TRANSPARENT);
//    for (Rect rect : rectArrayList) {
//      rectIV.drawUnselectedRect(rect);
//    }

    redrawrects();

    // updatelabelfile
    saveLabelFile();

  }

  /**
   * this method literally sets the selectedRect to null
   */
  public void removeSelectedRect(){
    this.selectedRect = null;
  }

  /**
   * this method will redraw the rects with unselected color
   */
  public void redrawrects(){
    rectBitmap.eraseColor(Color.TRANSPARENT);
    for (Rect rect : rectArrayList) {
      rectIV.drawUnselectedRect(rect);
    }
  }

  private void unselectAnyIfExist(){
    if(selectedRect==null){
      return;
    }

    removeSelectedRect();
    redrawrects();
    hideDeleteBtnCallback.run();

  }


  /***
   *
   *
   *
   *
   */

  public static class LabelDrawPadBuilder {

    private Callback drawBtnpressedcallback;
    private CallbackWithRect maskRectReadyCallback;
    private Runnable rectSelectedCallback;
    private Runnable hideDeleteBtnCallback;

    private LayoutInflater inflater;
    private ViewGroup container;
    private int position;
    private File imageFile;
    private CustomViewPager customViewPager;

    public LabelDrawPadBuilder(LayoutInflater inflater, ViewGroup container, int position) {
      this.inflater = inflater;
      this.container = container;
      this.position = position;
    }

    public LabelDrawPadBuilder setDrawBtnPressedCallback(Callback drawBtnpressedcallback) {
      this.drawBtnpressedcallback = drawBtnpressedcallback;
      return this;
    }

    public LabelDrawPadBuilder setMaskRectReadyCallback(CallbackWithRect maskRectReadyCallback) {
      this.maskRectReadyCallback = maskRectReadyCallback;
      return this;
    }

    public LabelDrawPadBuilder setImageFile(File imageFile) {
      this.imageFile = imageFile;
      return this;
    }

    public LabelDrawPadBuilder setCustomViewPager(CustomViewPager customViewPager) {
      this.customViewPager = customViewPager;
      return this;
    }

    public LabelDrawPadBuilder setRectSelectedCallback(Runnable rectSelectedCallback) {
      this.rectSelectedCallback = rectSelectedCallback;
      return this;
    }

    public LabelDrawPadBuilder setHideDeleteBtnCallback(Runnable hideDeleteBtnCallback) {
      this.hideDeleteBtnCallback = hideDeleteBtnCallback;
      return this;
    }

    public LabelDrawPad build() {
      return new LabelDrawPad(this);
    }



  }

}

File: app/src/main/java/com/example/chadrick/datalabeling/Models/LabelDrawPad.kt
package com.example.chadrick.datalabeling.Models

import android.graphics.*
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import com.example.chadrick.datalabeling.CustomComponents.*
import com.example.chadrick.datalabeling.CustomViewPager
import com.example.chadrick.datalabeling.R
import com.example.chadrick.datalabeling.Util
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.*
import java.util.*

/**
 * Created by chadrick on 18. 2. 12.
 */

class LabelDrawPad(inflater: LayoutInflater,
                   container: ViewGroup?,
                   drawBtnPressedCallback: () -> Boolean,
                   rectReadyCallback: (Rect) -> Unit,
                   imageFile: File,
                   customViewPager: CustomViewPager,
                   rectSelectedCallback: () -> Unit,
                   hideDeleteBtnCallback: () -> Unit
) {
    private val inflater = inflater
    private val container = container

    lateinit var rootview: View

    private val drawBtnPressedCallback = drawBtnPressedCallback

    lateinit var baseIV: DataImageImageView
    lateinit var rectIV: RenderedRectsImageView
    lateinit var drawIV: RectDrawImageView

    var image_width: Int = 0
    var image_height: Int = 0


    private val rectReadyCallback = rectReadyCallback

    private val customViewPager = customViewPager

    lateinit var mainCanvasPaint: Paint

    private val imageFile = imageFile
    lateinit var labelFile: File
    private var baseBitmap: Bitmap? = null
    private var rectBitmap: Bitmap? = null
    private val rectSelectedCallback = rectSelectedCallback
    private val hideDeleteBtnCallback = hideDeleteBtnCallback


    private val rectArrayList = ArrayList<Rect>()

    private val RECT_SELECT_AREA_PADDING = 50

    private var selectedRect: Rect? = null


    init {
        initElements()
    }


    private fun initElements() {

        rootview = inflater.inflate(R.layout.labeldrawpad_layout, container,
                false)
        baseIV = rootview.findViewById(R.id.baseiv)
        rectIV = rootview.findViewById(R.id.rectiv)
        drawIV = rootview.findViewById(R.id.drawiv)

        // need to register the IVs so that it can receive touch events
        val labelpadframelayout = rootview
                .findViewById<LabelPadFrameLayout>(R.id.labelpadframelayout)
        labelpadframelayout.baseIV = baseIV
        labelpadframelayout.maskIV = drawIV
        labelpadframelayout.rectIV = rectIV

        // create paint for maincanvas
        mainCanvasPaint = Paint()
        mainCanvasPaint.color = Color.rgb(255, 0, 0)
        mainCanvasPaint.strokeWidth = 5f
        mainCanvasPaint.style = Paint.Style.STROKE

        // dynamic width/height measure for drawIV
        val viewTreeObserver = drawIV.viewTreeObserver
        viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {

            override fun onGlobalLayout() {

                drawIV.passWH(drawIV.width, drawIV.height)
                drawIV.viewTreeObserver.removeOnGlobalLayoutListener(this)

            }
        })
        // create bitmap with imageFile and set up baseCanvas
        val options = BitmapFactory.Options()
        options.inPreferredConfig = Bitmap.Config.ARGB_8888
        options.inMutable = true
        val originalbitmap = BitmapFactory.decodeFile(imageFile.toString(), options)
        image_width = originalbitmap.width
        image_height = originalbitmap.height

        baseBitmap = Bitmap.createBitmap(originalbitmap.width, originalbitmap.height,
                Bitmap.Config.ARGB_8888)
        val basecanvas = Canvas(baseBitmap)
        basecanvas.drawBitmap(originalbitmap, 0f, 0f, null)
        // create blank bitmap
        rectBitmap = Bitmap
                .createBitmap(originalbitmap.width, originalbitmap.height, Bitmap.Config.ARGB_8888)
        val rectCanvas = Canvas(rectBitmap)
        // set baseIV
        baseIV.setImageBitmap(baseBitmap)
        baseIV.drawBtnPressedCallback = drawBtnPressedCallback


        // set rectIV
        rectIV.setImageBitmap(rectBitmap)
        rectIV.canvas = rectCanvas
        rectIV.customViewPager = customViewPager
        rectIV.addRectCallback = { rect -> rectArrayList.add(rect) }
        rectIV.saveLabelCallback = this::saveLabelFile
        rectIV.drawBtnpressedcallback = drawBtnPressedCallback
        rectIV.checkRectSelectCallback = this::checkRectSelect
        rectIV.unselectAnyIfExistCallback = this::unselectAnyIfExist

        drawIV.drawBtnpressedcallback = drawBtnPressedCallback
        drawIV.rectReadyCallback = rectReadyCallback

        // generate label json file based on image file
        val imagefilename = Util.getOnlyFilename(imageFile.name)
        val parentpath = imageFile.parent
        val printstr = parentpath + File.separator + imagefilename + ".json"
        labelFile = File(printstr)
        // check if labelfile exist. if exist, read it and draw them. if not exist, then do nothing.
        if (!labelFile.exists()) {
            // when json file doesn't exist

        } else {
            // if json file exists, then read it into a json object
            val readtext = StringBuilder()
            try {
                val bufferedReader = BufferedReader(FileReader(labelFile))
                var line: String?

                line = bufferedReader.readLine()

                while (line != null) {
                    readtext.append(line)
                    line = bufferedReader.readLine()
                }

            } catch (e: IOException) {
                e.printStackTrace()

            }

            val readresult = readtext.toString()
            try {
                val parsed = JSONObject(readresult)
                val bitmapfilename = parsed.getString("imgfile")
                val objects = parsed.getJSONArray("objects")
                for (i in 0 until objects.length()) {
                    val item = objects.getJSONObject(i)
                    val name = item.getString("name")
                    val rect = item.getJSONObject("rect")
                    val x1 = rect.getInt("x1")
                    val y1 = rect.getInt("y1")
                    val x2 = rect.getInt("x2")
                    val y2 = rect.getInt("y2")
                    val rectfromjson = Rect(x1, y1, x2, y2)
                    rectArrayList.add(rectfromjson)

                }

            } catch (e: JSONException) {
                e.printStackTrace()
            }

            // parsing the jsonfile finished.
            // now draw each rect in the maincanvas
            for (i in rectArrayList.indices) {
                val rectToDraw = rectArrayList.get(i)
                rectCanvas.drawRect(rectToDraw, mainCanvasPaint)
            }
            // invalidate to make sure that the new drawing are displayed
            rectIV.invalidate()

        }
    }


    private fun saveLabelFile(): Boolean {
        // when called, rewrite the labelfile with current info
        // first create a jsonobject and populate with current info
        val newRoot = JSONObject()
        try {
            newRoot.put("imgfile", imageFile.name)
            newRoot.put("w", image_width)
            newRoot.put("h", image_height)
            // create JSONarray object that contains all the current rects
            val rectarray = JSONArray()
            for (i in rectArrayList.indices) {
                val rect = rectArrayList[i]
                val `object` = JSONObject()
                `object`.put("name", "car")
                val jsonrect = JSONObject()
                jsonrect.put("x1", rect.left)
                jsonrect.put("y1", rect.top)
                jsonrect.put("x2", rect.right)
                jsonrect.put("y2", rect.bottom)
                `object`.put("rect", jsonrect)
                rectarray.put(`object`)
            }
            newRoot.put("objects", rectarray)

        } catch (e: JSONException) {
            e.printStackTrace()
            return false
        }

        // newroot is ready. write it to labelFile
        try {
            val bufferedWriter = BufferedWriter(FileWriter(labelFile))
            bufferedWriter.write(newRoot.toString())
            bufferedWriter.flush()
            bufferedWriter.close()
        } catch (e: IOException) {
            e.printStackTrace()
            return false
        }

        // finished overwriting the labelfile
        // job finished
        // return true since we have successfully written it.
        return true

    }

    fun enableTouches() {

        baseIV.setTouchEnable(true)
        rectIV.setTouchEnable(true)
        drawIV.TouchEnabled = true
        //    drawIV.setTouchEnable(true);
    }

    fun disableTouches() {

        baseIV.setTouchEnable(false)
        rectIV.setTouchEnable(false)
        drawIV.TouchEnabled = false
        //    drawIV.setTouchEnable(false);
    }

    private fun checkRectSelect(point: Point): Boolean {
        // check with all rects. the first hit will be set as selected and
        // this method will return true
        // if there are no hits, then return false
        val touch_x = point.x
        val touch_y = point.y
        for (rect in rectArrayList) {
            val biggerRect = Rect(rect.left - RECT_SELECT_AREA_PADDING,
                    rect.top - RECT_SELECT_AREA_PADDING,
                    rect.right + RECT_SELECT_AREA_PADDING,
                    rect.bottom + RECT_SELECT_AREA_PADDING)
            val smallerRect = Rect(rect.left + RECT_SELECT_AREA_PADDING,
                    rect.top + RECT_SELECT_AREA_PADDING,
                    rect.right - RECT_SELECT_AREA_PADDING,
                    rect.bottom - RECT_SELECT_AREA_PADDING)
            // assume that there are no conflicts with the sizes of the bigger and smaller rect
            var biggercontain = false
            var smallercontain = false
            if (biggerRect.contains(touch_x, touch_y)) {
                biggercontain = true
            }
            if (smallerRect.contains(touch_x, touch_y)) {
                smallercontain = true
            }
            if (biggercontain  && !smallercontain) {
                // okay so a hit was found, and we know which rectangle it is.
                // we need to redraw it, and it will eventually be done with
                // rectIV
                // the rectangle of interest is already drawn.
                // instead of redrawing the whole thing(including othe rectangles),
                // why not just overdraw the rectangle of interest with anther paint?
                // first draw unselected rect of the previous existing selectedRect

                selectedRect?.let { rectIV.drawUnselectedRect(it) }
                selectedRect = rect

                selectedRect?.let { rectIV.drawSelectedRect(it) }
                // display delete btn
                rectSelectedCallback()
                return true
            }

        }
        return false
    }

    fun drawRect() {

        rectIV.drawRect()
    }

    fun eraseDrawnRect() {

        drawIV.eraseall()
    }

    /**
     * this will delete the selectedRect from the list
     * and redraw canvas without that rect
     */
    fun deleteSelectedRect() {
        // delete the selectedrect
        if (selectedRect == null) {
            return
        }
        // remove it from rectarray
        selectedRect?.let { rectArrayList.remove(it) }
        // nullify the selectedrect variable
        selectedRect = null
        redrawrects()
        // updatelabelfile
        saveLabelFile()

    }

    /**
     * this method literally sets the selectedRect to null
     */
    fun removeSelectedRect() {

        this.selectedRect = null
    }

    /**
     * this method will redraw the rects with unselected color
     */
    fun redrawrects() {

        rectBitmap?.eraseColor(Color.TRANSPARENT)
        for (rect in rectArrayList) {
            rectIV.drawUnselectedRect(rect)
        }
    }

    private fun unselectAnyIfExist() {

        if (selectedRect == null) {
            return
        }
        removeSelectedRect()
        redrawrects()
        hideDeleteBtnCallback()

    }

}